README.md
# yaccl
yet another ChemClassifier (Python based on wikibase-cli and rdkit)

This is a proof of the concept of a simple compound classifier that relies completely on knowledge from Wikidata. At this stage it uses InChI keys, SMILES and SMARTS strings, downloaded from Wikidata, to get hits on compounds or classes. About 650 biosynthetic processes from Gene Ontology are directly associated with classes and compounds, so they are potential hits. Matching is done by going through the list of classes---fast enough to find all hits in the dataset with >110k classes within a few seconds.

While yaccl is, in principle, a general classifier, development of patterns focuses on biomolecules.

## Version / Progress

The current version is 2122.

- [X] terpenoids
- [X] steroids
- [X] flavonoids
- [X] alkaloids (25% unspecified but recognized)
- [X] polyketides
- [ ] fatty acyls (in progress)


## Prerequisites

* Python 3
* rdkit python package
* wikibase-cli (optional, for dataset updates)

Cloning the repo will get you all datasets, so no need to download from Wikidata if you just want to play with yaccl.

## Installation

Clone the repo.

## Usage

Change into the repo `src` directory. There are four scripts. One is for classification, and three are for updating / statistics of datasets.

### classify

Takes a SMILES/InChI and outputs a classification. Examples:

```
$ python3 classify.py -d ./ -m 'CCO'
reading biosyn data
reading class pattern data
reading superclass data
collecting hits...
purging redundant hits
('Q153', 'ethanol', 'LFQSCWFLJHTTHZ-UHFFFAOYSA-N') {('GO:0044576', 'pentose catabolic process to ethanol'), ('GO:0044577', 'xylose catabolic process to ethanol'), 'GO:0006115', 'ethanol biosynthetic process', ('GO:0019431', 'acetyl-CoA biosynthetic process from ethanol')}
('Q2832210', 'primary alcohol', '[OX2H1][CX4H2]') {'primary alcohol biosynthetic process', 'GO:0034309'}
```
```
$ python3 classify.py -d ./ -m 'InChI=1S/C15H22O10/c16-3-6-9(19)10(20)11(21)14(23-6)24-13-7-5(1-2-22-13)8(18)12-15(7,4-17)25-12/h1-2,5-14,16-21H,3-4H2/t5-,6-,7-,8+,9-,10+,11-,12+,13+,14+,15-/m1/s1'
reading biosyn data
reading class pattern data
reading superclass data
collecting hits...
purging redundant hits
('Q105151716', '2-[[5-hydroxy-2-(hydroxymethyl)-3,9-dioxatricyclo[4.4.0.02,4]dec-7-en-10-yl]oxy]-6-(hydroxymethyl)oxane-3,4,5-triol', 'OCC1OC(OC2OC=CC3C(O)C4OC4(CO)C23)C(O)C(O)C1O') {'GO:0016138', 'glycoside biosynthetic process'}
('Q408028', 'epoxide', 'C1OC1') {'GO:1901503', 'ether biosynthetic process'}
('Q2832210', 'primary alcohol', '[OX2H1][CX4H2]') {'primary alcohol biosynthetic process', 'GO:0034309'}
('Q2832211', 'secondary alcohol', '[CX4H1][OX2H1]') {'GO:1902653', 'secondary alcohol biosynthetic process'}
('Q74173050', 'D-glucoside', '[C@@H]1(OC([C@H](O)[C@H]([C@@H]1O)O)O*)CO') {'GO:0016138', 'glycoside biosynthetic process'}
('Q74568405', 'alpha-D-mannoside', 'OC[C@@H]1[C@H]([C@@H]([C@H](O)[C@H](O1)O*)O)O') {'GO:0016138', 'glycoside biosynthetic process'}
('Q75078202', 'pyranoside', 'O1C(C(C(C(C1O*)O)O)O)CO') {'GO:0016138', 'glycoside biosynthetic process'}
('Q75082510', 'beta-D-galactoside', 'OC[C@H]1O[C@@H](O[*])[C@H](O)[C@@H](O)[C@H]1O') {'GO:0016138', 'glycoside biosynthetic process'}
('Q105151716', '2-[[5-hydroxy-2-(hydroxymethyl)-3,9-dioxatricyclo[4.4.0.02,4]dec-7-en-10-yl]oxy]-6-(hydroxymethyl)oxane-3,4,5-triol', 'OCC1OC(OC2OC=CC3C(O)C4OC4(CO)C23)C(O)C(O)C1O') {'GO:0016138', 'glycoside biosynthetic process'}
```

The `-j` option provides JSON format. This does only work if a natural product is recognized. Then, it gives the full path to root in the ontology:
```
$ python3 classify.py -d ./ -m 'InChI=1S/C28H32O15/c1-39-14-7-15-18(12(32)6-13(40-15)10-2-4-11(31)5-3-10)22(35)19(14)26-27(24(37)21(34)16(8-29)41-26)43-28-25(38)23(36)20(33)17(9-30)42-28/h2-7,16-17,20-21,23-31,33-38H,8-9H2,1H3/t16-,17-,20-,21-,23+,24+,25-,26+,27-,28+/m1/s1' -j 
[{"classification_names": ["flavonoid", "2-phenylchromane flavonoid", "flavone", "6C-substituted flavone", "6C-glycosylated flavone", "spinosin"], "biological_process": ["flavone biosynthetic process", "GO:0051553"]}]
```
 
Given a file of InChI strings the `-t` option runs a test checking that at least one of the hits is a subclass of the test class item.
```
$ python3 classify.py -t test/sesterterpenoids.txt -d ./ 
reading biosyn data
reading class pattern data
reading superclass data
classifying testfile test/sesterterpenoids.txt
test class: Q107363222
1 OK
2 OK
3 OK
4 OK
5 OK
6 OK
7 OK
8 OK
9 FAIL: InChI=1S/C25H44/c1-9-21(6)13-11-17-25(8,16-10-12-19(2)3)24-18-22(7)14-15-23(24)20(4)5/h9,18-19,21,23-24H,1,4,10-17H2,2-3,5-8H3
...
```

### datasets
The difference between the `biosyn` and the `class-pattern` datasets is that the former contains classes AND compounds that have associated GO processes in Wikidata (possibly lacking any pattern), while the latter contains all classes (no compounds) that have associated patterns (SMILES, SMARTS or InChI).

The design of datasets will be subject to change in future versions.

Using the `-q` option with the dataset scripts depends on a working `wikibase-cli` installation
to attempt an update of the respective database, directly from Wikidata. This is not necessary
if you frequently `git pull` this repo, as any Wikidata changes will be reflected by repo updates.

### biosyn-class-stats
Output stats of `biosyn-classes` dataset.

```
$ python3 biosyn-class-stats.py 
#items: 1167
#InChI keys: 635
#InChI1 keys: 242
#smarts: 42
#smiles: 694
# w/o pattern: 436
```

### class-pattern
Output stats of `class-pattern` dataset.

```
$ python3 class-pattern.py 
#items: 113788
#InChI keys: 112432
#InChI1 keys: 81120
#smarts: 374
#smiles: 1024
#wildcard smiles: 979
```
Yes, Wikidata has >110k compound classes, most of which (>81k) are compounds with unspecified stereochemistry, i.e. they are groups of stereoisomers.

### Contribute

release-notes.md
===2101===
* initial version

===2102===
* patterns: added >30 SMARTS, replacing unsuitable wildcard SMILES
* biosyn: added 24 SMARTS
* classify: add option to load superclass dataset; does nothing atm
* add `class-subclass.py` for updating superclass dataset

===2103===
* unified dataset handling (all datasets are loaded from given directory)
* data update

===2104===
* classify: remove redundant hits
* classify: get nearest parent GO process if none on class
* classify: add option to get redundant hits as before
* superclass data: more class types, bigger ontology, dead-end diagnostic

===2105===
* all data: monoterpenoids
* test data: monoterpenoids, 14/197 FAIL
* classify: test capability
* multiple SMARTS (denoting sets of patterns) now handled

===2106===
* all data: sesquiterpenoids
* test data: sesquiterpenoids, 7/240 FAIL

===2107===
* all data: diterpenoids
* test data: diterpenoids, 0/177 FAIL
* don't load SMILES/SMARTS if InChI present, 10x speed up of classification

===2108===
* remove bogus pattern leading to Boost warnings
* revised/added to acyclic monoterpenoids, FAIL: 0/197
* all data: sesterterpenoids
* test data: sesterterpenoids, 1/14 FAIL
* split pattern query, makes it faster, more resilient, one third reduction in pattern db size

===2109===
* some new files were missing in last release
* reduce test run output
* all data: triterpenoids
* test data: triterpenoids, 1/85 FAIL

===2110===
* pattern fixes (not all possible double bonds were allowed)
* all data: tetraterpenoids
* test data: tetraterpenoids, 7/199 FAIL

===2111===
* all data: steroids
* test data: steroids, 3/1561 FAIL

===2112===
* add -n option to classify small sample of WD natural products
* all data: flavonoids pt. 1 (est. 60% of all)
* test data: flavonoids

===2113===
* all data: flavonoids pt. 2 (est. 85% of all)

===2114===
* add -j option for JSON output (NP only)
* all data: flavonoids pt. 3 (est. 95% of all)

==2115==
* all data: alkaloids pt. 1 (est. 25% of all)
* test data: alkaloids

==2116==
* fixed bug where compounds with biosyn process were not seen with -j option
* all data: alkaloids pt. 2 (est. 45% of all)

==2117==
* sort pattern data on SMARTS hash to avoid most reshuffling
* all data: alkaloids pt. 3 (est. 62% of all)

==2118==
* fix bug where hits to NP root classes were ignored
* again fix data write order; Py hash() was salted
* add rule to recognize unspecified alkaloids
* all data: alkaloids pt. 4 (est. 35% unspecified)
* add option ouputting NP roots in JSON format
* revamp top NP hierarchy

==2119==
* all data: polyketides pt. 1 (est. 90% of all)
* test data: polyketides
* prettify `--list_nproots`
* further NP root optimizations

==2120==
* test data polyketides was missing
* add rule to recognize unspecified macrolides
* all data: polyketides pt. 2 (est. 97% of all)

==2121==
* classify: revamp -n code
* classify: include mol data with JSON output
* improve macrolide rule
* reduced unspec. alkaloids to 25% by adding more small groups
* unspec. alkaloids and makrolides tagged in WD


src/classify.py
import csv, os, json, argparse, sys, random
from rdkit import Chem

"""
Load data from directory, classify given molecule or test file 
"""

# Walk reverse ontology subgraph (root = hitem) upwards, depth first. If any member of hitemset
# is encountered, add it to minor set. These are later discarded, leaving only the most
# specialized hits
def walk_ont(sitems, hitemset, hitem, minors):
    ch = sitems.get(hitem)
    if ch is None:
        return
    for c in ch:
        if c in hitemset:
            minors.add(c)
        walk_ont(sitems, hitemset, c, minors)

# Walk reverse ontology subgraph (root = node) upwards, breadth first. The first encountered
# GO process is returned.
def walk_ont_go(sitems, node, gos):
    visited = set()    # List to keep track of visited nodes.
    queue = []      #Initialize a queue
    visited.add(node)
    queue.append(node)

    while queue:
        s = queue.pop(0)
        go = gos.get(s)
        if go is not None:
            return go
        ch = sitems.get(s)
        if ch is None:
            continue
        for c in ch:
            if c not in visited:
                visited.add(c)
                queue.append(c)
    return None

# Walk reverse ontology subgraph (root = hitem) upwards, depth first.
# Return True if citem is encountered (i.e. citem is superclass of hitem)
def walk_find(sitems, hitem, citem):
    ch = sitems.get(hitem)
    # print('{} {} {}'.format(hitem, citem, ch))
    if ch is None:
        return False
    if citem in ch:
        return True
    for c in ch:
        if walk_find(sitems, c, citem):
            return True
    return False

# Walk reverse ontology subgraph (root = hitem) upwards, depth first.
# If one of items in list is encountered return full path to it 
path = []
def walk_find_list(sitems, hitem, item_set):
    global path
    ch = sitems.get(hitem)
    # print('{} {} {}'.format(hitem, citem, ch))
    if ch is None:
        return False
    for c in ch:
        if c is None:
            continue
        if c in item_set:
            path.append(str(c))
            return True
        if walk_find_list(sitems, c, item_set):
            path.append(c)
            return True
    return False

# Find path to topmost natural product, return as array of names
def path_to_nproot(hit, nplist):
    global path
    if hit in nplist:
        return [hit]
    path = []
    success = walk_find_list(sitems, hit, set(nplist))
    if not success:
        return []
    path.append(hit)
    return path

ctr = 0
def is_elem(e, s):
    global ctr
    ctr = ctr + 1
    return e in s

# Walk ontology downwards, breadth first, starting from given root, add children to set
def walk_collect_npclasses(edges, node, collection):
    visited = set()    # List to keep track of visited nodes.
    queue = []      #Initialize a queue
    visited.add(node)
    queue.append(node)

    while queue:
        s = queue.pop(0)
        e = edges.get(s)
        if e is None:
            continue
        for c in e:
            if c not in visited:
                visited.add(c)
                queue.append(c)
    collection |= visited

#rule A-1
def check_unspec_alkaloid(mol):
    #ring-N
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[#7R]')):
        return False
    #!peptide
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[$([NR0]),$([Nr;!r3;!r4;!r5;!r6;!r7;!r8])]~C(~[OX1H0,OX2H1])')):
        return False
    #!tetrapyrroles
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[#6]1~[#6]~[#6]2~[#6]~[#6]3~[#6]~[#6]~[#6](~[#6]~[#6]4~[#6]~[#6]~[#6](~[#6]~[#6]5~[#6]~[#6]~[#6](~[#6]~[#6]~1~[#7]~2)~[#7]~5)~[#7]~4)~[#7]~3')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[#6]1~[#6]~[#7]~[#6](~[#6R0]~[#6]2~[#6]~[#6]~[#6](~[#6R0]~[#6]3~[#6]~[#6]~[#6](~[#6R0]~[#6]4~[#6]~[#6]~[#6]~[#7]~4)~[#7]~3)~[#7]~2)~[#6]~1')):
        return False
    #!nucleotides
    if mol.HasSubstructMatch(Chem.MolFromSmarts('POCC1CCC(n2cnc3cncnc32)O1')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('POCC2CCC([#7R1]1~[#6R1]~[#7R1]~[#6R1]~[#6R1]~[#6R1]~1)O2')):
        return False
    #!diketopiperazine
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[#6]1~[#7]~[#6](~O)~[#6]~[#7]~[#6](~O)~1')):
        return False
    return True

#rule H-1
def check_unspec_hydrocarbon(mol):
    # aliphatic
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('C')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[c,#7,#8,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    return True

def check_acyclic_ketone(mol):
    # aliphatic
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('CC(=O)C')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('O~CC(~O)C~O')):
        return False
    return True

def check_acyclic_aldehyde(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('O=[CX3H1]C')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[$(O);!$(O=C(C)C);!$(O=[CX3H1]C)]')):
        return False
    return True

def check_acyclic_epoxide(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[OX2H0]1[CR1][CR1]1')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[C;r{4-}]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    return True

def check_acyclic_ether(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('O([$(C);!$(C(~O)O)])[$(C);!$(C(~O)O)]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    return True

def check_acyclic_ester(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('C[CX3](=O)[OX2H0][$(C);!$(C=O)]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    return True

def check_acyclic_alcohol(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[$(C);!$(C~O)]~[$(C);!$(C=O)][OX2H1]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('OCC(O)C~O')):
        return False
    return True

def check_acyclic_peroxide(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[OX2][OX2]')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,#16,F,Cl,Br,I,P,As,Se]')):
        return False
    return True

def check_fatty_acid_anion(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[CX3](=O)[OX1H0-]')):
        return False
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('C~C~C~C')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,S,P,As]')):
        return False
    return True

def check_fatty_acid(mol):
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[CX3](=O)[OX2H1]')):
        return False
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('C~C~C~C')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR,c,#7,o,S,P,As]')):
        return False
    return True

#rule M-1
def check_unspec_macrolide(mol):
    #8+ macro lactone
    if not mol.HasSubstructMatch(Chem.MolFromSmarts('[OR1;r{8-}][CR1;r{8-}](=O)')):
        return False
    #!macrocyclic benzene
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[cR]([C,O;r{9-}])1[cR]([#6,O])[cR1][cR1][cR1][cR1]1')):
        return False
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[cR]([C,O;r{9-}])1[cR][cR]([#6,O])[cR1][cR1][cR1]1')):
        return False
    #!cyclopeptide
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[NR]~[CR](~[OX1H0,OX2H1])[CR][NR;r{7-}]~[CR](~[OX1H0,OX2H1])')):
        return False
    #!cyclodepsipeptide
    if mol.HasSubstructMatch(Chem.MolFromSmarts('[CR](=O)[NR][CR][CR](=O)[OR][CR][CR](=O)[NR][CR][CR](=O)[OR]')):
        return False
    return True

# Try to match ALL patterns. Remove redundant. Return remaining.
def get_hits(mol, silent=False):
    if not silent:
        print('collecting hits...')
    hits = {}
    # match the InChI keys
    ik = Chem.MolToInchiKey(mol)
    it = iks.get(ik)
    if it is not None:
        go = ''
        g = gos.get(it)
        if g is not None:
            go = g
        hits[it] = (it, labels.get(it), ik)
        if args.rawhits:
            print('{} {} {} {}'.format(it, labels.get(it), ik, go))
    else:
        it = ik1s.get(ik[:14])
        if it is not None:
            go = ''
            g = gos.get(it)
            if g is not None:
                go = g
            hits[it] = (it, labels.get(it), ik[:14])
            if args.rawhits:
                print('{} {} {} {}'.format(it, labels.get(it), ik[:14], go))

    ps_default = Chem.AdjustQueryParameters()
    ps_default.adjustDegreeFlags = Chem.AdjustQueryWhichFlags.ADJUST_IGNOREDUMMIES | Chem.AdjustQueryWhichFlags.ADJUST_IGNORECHAINS
    ps_ignoreDummies = Chem.AdjustQueryParameters()
    ps_ignoreDummies.adjustDegreeFlags = Chem.AdjustQueryWhichFlags.ADJUST_IGNOREDUMMIES
    for it,itsuplist in sitems.items():
        sm = smiles.get(it)
        smas = smarts.get(it)
        if sm is None and smas is None:
            continue
        if (not (sm is None or len(sm) == 0)) and sm.find('*') < 0:
            if sm.find('@') >= 0:
                continue
        # this class has a non-wildcard SMILES
            try:
                q = Chem.MolFromSmiles(sm)
                pats = [Chem.AdjustQueryProperties(q, ps_ignoreDummies)]
            except Exception:
                continue
        elif not smas is None:
        # this class has a wildcard SMILES or a SMARTS
            #print('---{} {} {}'.format(it, labels.get(it), sma))
            try:
                pats = [Chem.MolFromSmarts(sma) for sma in smas]
                #pat = Chem.AdjustQueryProperties(pat, ps_default)
                #if it == 'Q12748271':
                #    print(mol.HasSubstructMatch(pat))
                #    exit()
                #print('====={} {}'.format(it, sma))
            except Exception:
                continue
        else:
            try:
                q = Chem.MolFromSmiles(sm)
                pats = [Chem.AdjustQueryProperties(q, ps_ignoreDummies)]
            except Exception:
                continue
        
        if pats is None:
            print('========defect pattern: {} {}'.format(it, sma))
            continue
        for p in pats:
            if mol.HasSubstructMatch(p):
                go = ''
                g = gos.get(it)
                if g is not None:
                    go = g
                hits[it] = (it, labels.get(it), sm)
                if args.rawhits:
                    print('{} {} {} {}'.format(it, labels.get(it), sm, go))
                break

    if check_unspec_hydrocarbon(mol):
        hits['Q109910560'] = ('Q109910560', 'unspecified hydrocarbon', 'Rule H-1')
    else:
        if check_fatty_acid_anion(mol):
            hits['Q71653081'] = ('Q71653081', 'fatty acid anion', 'Rule F-8')
        if check_fatty_acid(mol):
            hits['Q61476'] = ('Q61476', 'fatty acid', 'Rule F-9')
        if check_acyclic_aldehyde(mol):
            hits['Q109923365'] = ('Q109923365', 'acyclic aldehyde', 'Rule F-2')
        if check_acyclic_ether(mol):
            hits['Q109923862'] = ('Q109923862', 'acyclic ether', 'Rule F-4')
        if check_acyclic_epoxide(mol):
            hits['Q109923685'] = ('Q109923685', 'acyclic epoxide', 'Rule F-3')
        if check_acyclic_ketone(mol):
            hits['Q109911294'] = ('Q109911294', 'acyclic ketone', 'Rule F-1')
        if check_acyclic_ester(mol):
            hits['Q109923912'] = ('Q109923912', 'acyclic ester', 'Rule F-5')
        if check_acyclic_alcohol(mol):
            hits['Q378871'] = ('Q378871', 'fatty alcohol', 'Rule F-6')
        if check_acyclic_peroxide(mol):
            hits['Q109946855'] = ('Q109946855', 'acyclic peroxide', 'Rule F-7')
        if check_unspec_alkaloid(mol):
            hits['Q70702'] = ('Q70702', 'unspecified alkaloid', 'Rule A-1')
        if check_unspec_macrolide(mol):
            hits['Q422687'] = ('Q422687', 'unspecified macrolide', 'Rule M-1')
    
    if not silent:
        print('purging redundant hits')
    hitemset = set(hits.keys())
    minors = set()
    for hit in hitemset:
        walk_ont(sitems, hitemset, hit, minors)
    for m in minors:
        hits.pop(m)

    return hits


# Initiate the parser
parser = argparse.ArgumentParser()
parser.add_argument("-d", "--data", help="data directory", required=True)
parser.add_argument("-m", "--molecule", help="molecule to classify (SMILES/InChi)")
parser.add_argument("-a", "--rawhits", help="output all hits before hit processing starts",
        action="store_true")
parser.add_argument("-t", "--testfile", help="classify all InChis from file, first line may contain check item", type=str)
parser.add_argument("-n", "--nptest", help="Load N random NP Inchis from WD and classify", action="store_true")
parser.add_argument("-j", "--json", help="together with -m outputs JSON formatted result", action="store_true")
parser.add_argument("-N", "--natural", help="prune ontology to only include natural products", action="store_true")
parser.add_argument("--list_nproots", help="list of topmost natural products classes", action="store_true")

# Read arguments from the command line
args = parser.parse_args()
is_normal_run = not (args.molecule is None) or not(args.testfile is None) or args.nptest
is_extra_service = args.list_nproots

if not is_normal_run and not is_extra_service:
    print('One of -m or -t or -n is needed')
    parser.print_usage()
    exit()

silent = args.json

if not silent:
    print(args)

if is_normal_run and args.testfile is None and args.nptest is False:
    if args.molecule.find('InChI=') >= 0:
        mol = Chem.MolFromInchi(args.molecule)
    else:
        mol = Chem.MolFromSmiles(args.molecule)


iks = {}
ik1s = {}
labels = {}
smiles = {}
smarts = {}

# read bio process data
with open(args.data + 'data-biosyn-classes.json', 'r') as f:
    if not silent:
        print('reading biosyn class data')
    s = f.read()
    jol = json.loads(s)

gos = {}
for d in jol:
    dd = d.get('item')
    it = dd.get('value')
    ik = d.get('p235')
    if ik is not None:
        iks[ik] = it
        if ik[15:25] == 'UHFFFAOYSA':
            ik1s[ik[:14]] = it
    else:
        sm = None
        p233 = None
        p2017 = None
        p233 = d.get('p233')
        p2017 = d.get('p2017')
        if p2017 is not None and len(p2017) > 0:
            sm = p2017
        elif p233 is not None and len(p233) > 0:
            sm = p233
        if sm is not None:
            smiles[it] = sm
        p8533 = d.get('p8533')
        if p8533 is not None and len(p8533) > 0:
            t = smarts.get(it)
            if t is None:
                smarts[it] = [p8533]
            else:
                t.append(p8533)
    g = gos.get(it)
    gotup = (d.get('goid'), d.get('goLabel'))
    if g is not None:
        g.add(gotup)
        continue
    else:
        gos[it] = set(gotup)
    lab = dd.get('label')
    labels[it] = lab

sitems = {}
with open(args.data + 'data-biosyn-compounds.json', 'r') as f:
    if not silent:
        print('reading biosyn compound data')
    s = f.read()
    jol = json.loads(s)

for d in jol:
    dd = d.get('item')
    it = dd.get('value')
    ik = d.get('p235')
    if ik is not None:
        iks[ik] = it
        if ik[15:25] == 'UHFFFAOYSA':
            ik1s[ik[:14]] = it
    else:
        sm = None
        p233 = None
        p2017 = None
        p233 = d.get('p233')
        p2017 = d.get('p2017')
        if p2017 is not None and len(p2017) > 0:
            sm = p2017
        elif p233 is not None and len(p233) > 0:
            sm = p233
        if sm is not None:
            smiles[it] = sm
        p8533 = d.get('p8533')
        if p8533 is not None and len(p8533) > 0:
            t = smarts.get(it)
            if t is None:
                smarts[it] = [p8533]
            else:
                t.append(p8533)
    g = gos.get(it)
    gotup = (d.get('goid'), d.get('goLabel'))
    if g is not None:
        g.add(gotup)
        continue
    else:
        gos[it] = set(gotup)
    lab = dd.get('label')
    labels[it] = lab
    p31 = d.get('p31')
    if p31 is not None:
        sitems[it] = p31

# read pattern data
with open(args.data + 'data-class-pattern.json', 'r') as f:
    if not silent:
        print('reading class pattern data')
    s = f.read()
    jol = json.loads(s)

for d in jol:
    dd = d.get('item')
    it = dd.get('value')
    g = gos.get(it)
    if g is not None:
        continue
    lab = dd.get('label')
    ik = d.get('p235')
    if ik is not None:
        iks[ik] = it
        if ik[15:25] == 'UHFFFAOYSA':
            ik1s[ik[:14]] = it
    else:
        sm = None
        p233 = None
        p2017 = None
        p233 = d.get('p233')
        p2017 = d.get('p2017')
        if p2017 is not None and len(p2017) > 0:
            sm = p2017
        elif p233 is not None and len(p233) > 0:
            sm = p233
        if sm is not None:
            smiles[it] = sm
        p8533 = d.get('p8533')
        if p8533 is not None and len(p8533) > 0:
            t = smarts.get(it)
            if t is None:
                smarts[it] = [p8533]
            else:
                t.append(p8533)
    labels[it] = lab

childs = {}
# filling subclass structure
with open(args.data + 'data-class-subclass.json', 'r') as f:
    if not silent:
        print('reading superclass data')
    s = f.read()
    jol = json.loads(s)

for d in jol:
    it = d.get('item')
    sup = d.get('super')
    if not args.natural:
        i = sitems.get(it)
        if i is not None:
            i.append(sup)
        else:
            sitems[it] = [sup]
    c = childs.get(sup)
    if c is not None:
        c.append(it)
    else:
        childs[sup] = [it]

with open(args.data + 'data-class-subclass-names.json', 'r') as ff:
    s = ff.read()
    jol = json.loads(s)

for d in jol:
    it = d.get('item')
    lab = d.get('lab')
    if lab is None:
        continue
    labels[it] = lab

# root items missing from query data
sitems['Q2393187'] = ['Q43460564']
labels['Q2393187'] = 'molecular entity'
labels['Q43460564'] = 'chemical entity'

#load list of NP roots
nplist = set()
with open('natural.txt', 'r') as nf:
    nl = nf.readlines()
    nplist = set([line.rstrip() for line in nl])

if args.list_nproots:
    j = []
    for i in nplist:
        d = {}
        d['item'] = i
        d['name'] = labels.get(i)
        j.append(d)
    print(json.dumps(j, indent=0))
    exit()

#create specialized ontology
if args.natural:
    nps = set()
    for nproot in nplist:
        walk_collect_npclasses(childs, nproot, nps)
#        print((nproot,len(nps),ctr))
        ctr = 0
    for np in nps:
        clist = childs.get(np)
        if clist is None:
            continue
        for c in clist:
            i = sitems.get(c)
            if i is not None:
                i.append(np)
            else:
                sitems[c] = [np]

if args.testfile is not None:
    print('classifying testfile {}'.format(args.testfile))
    with open(args.testfile, 'r') as tf:
        tfs = [line.strip() for line in tf.readlines()]
    count = 0
    fcount = 0
    for t_inchi in tfs:
        if t_inchi.startswith('Q'):
            check_item = t_inchi
            print('test class: {}'.format(check_item))
            continue
        mol = Chem.MolFromInchi(t_inchi)
        if mol is None:
            print('{} {}'.format(count, t_inchi))
            exit()
        hits = get_hits(mol, silent=True)
        count = count + 1
        if hits.get(check_item) is None and all([walk_find(sitems, hit, check_item) == False for hit in hits.keys()]):
            print('{} FAIL: {}'.format(count, t_inchi))
            fcount = fcount + 1
        else:
            print('{} OK'.format(count, hits))
    print('FAIL: {}/{}'.format(fcount, count))
elif args.nptest:
    N = 100
    print('classifying {} random small biomolecules'.format(N))
    with open(args.data + 'test/npitems.txt', 'r') as npf:
        qs = [line.strip() for line in npf.readlines()]
    sample = random.choices(qs, k=N)
    query="""
    SELECT ?inchi
    WHERE
    {{
      VALUES ?item {{ {} }}
      ?item wdt:P234 ?inchi.
    }}
    """.format('wd:' + ' wd:'.join(sample))
    f = open('npsample.rq', 'w')
    f.write(query)
    f.close()
    ret = os.popen('wd sparql npsample.rq >npsample.json')
    if ret.close() is not None:
        print('query failed')
        exit()
    with open('npsample.json', 'r') as tf:
        tfs = [line.strip() for line in tf.readlines()]
    count = 0
    fcount = 0
    j = []
    for t_inchi in tfs:
        mol = Chem.MolFromInchi(t_inchi)
        if mol is None:
            print('{} {}'.format(count, t_inchi))
            exit()
        hits = get_hits(mol, silent=True)
        count = count + 1
        found = False
        for hit in hits.keys():
            p = path_to_nproot(hit, nplist)
            if len(p) > 0:
                if args.json:
                    d = { 'result': True,
                            'classification_names': [labels.get(it) for it in p]
                        }
                    j.append(d)
                else:
                    print('{} OK {}'.format(count, labels.get(p[-1])))
                found = True
                break
        if not found:
            if args.json:
                d = { 'result': False,
                        'InChi': t_inchi,
                    }
                j.append(d)
            else:
                print('{} FAIL: {}'.format(count, t_inchi))
            fcount = fcount + 1
    if args.json:
        print(json.dumps(j))
    else:
        print('FAIL: {}/{}'.format(fcount, count))
else:
    silent = False
    if args.json:
        silent = True
    hits = get_hits(mol, silent)
    if args.rawhits:
        print('----------------------------------')
    D = {}
    D['molecule'] = args.molecule
    D['ikey'] = Chem.MolToInchiKey(mol)
    j = []
    for hit in hits.keys():
        if args.json:
            p = path_to_nproot(hit, nplist)
            if len(p) == 0:
                continue
            d = {}
            d['classification_names'] = [labels.get(it) for it in p]
        #finding closest biosynthetic processes
        cgo = walk_ont_go(sitems, hit, gos)
        if args.json:
            if cgo is not None:
                d['biological_process'] = list(cgo)
            j.append(d)
        else:
            print('{} {}'.format(hits.get(hit), cgo))
    if args.json:
        if len(j) > 0:
            D['hits'] = j
        print(json.dumps(D))


